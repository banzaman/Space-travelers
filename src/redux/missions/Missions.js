import fetchMission from './api';

const GET_MISSION = 'MISSIONS_DISPLAYED';

const initialState = [];

export default function missionsReducer(state = initialState, action) {
  switch (action.type) {
    case GET_MISSION: return action.payload;
    default: return state;
  }
}

export const showMissions = () => async (dispatch) => {
  let missions;
  try {
    missions = await fetchMission();
    dispatch({
      type: GET_MISSION,
      payload: missions,
    });
  } catch (err) {
    return err;
  }
  return missions;
};
