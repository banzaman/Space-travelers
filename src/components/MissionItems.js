import { Badge, Button } from 'react-bootstrap';
import { useDispatch } from 'react-redux';
import PropTypes from 'prop-types';
import { MemberShip } from '../redux/missions/missionsSlice';
import '../assets/css/Mission.css';

export default function MissionItems({
  id, name, description, reserved,
}) {
  const dispatch = useDispatch();
  if (!reserved) {
    return (
      <li key={id} className="mission-list">
        <h5 className="mission-name">{name}</h5>
        <p className="description">{description}</p>
        <span className="status"><Badge bg="secondary">NOT A MEMBER</Badge></span>
        <div className="join">
          <Button
            variant="outline-secondary"
            onClick={() => dispatch(MemberShip(id))}
          >
            Join Mission
          </Button>
        </div>
      </li>
    );
  }
  return (
    <li key={id} className="mission-list">
      <h5 className="mission-name">{name}</h5>
      <p className="description">{description}</p>
      <span className="status"><Badge bg="primary">Active Member</Badge></span>
      <div className="join">
        <Button
          variant="outline-danger"
          onClick={() => dispatch(MemberShip(id))}
        >
          Leave Mission
        </Button>
      </div>
    </li>
  );
}

MissionItems.propTypes = { id: PropTypes.string.isRequired };
MissionItems.propTypes = { name: PropTypes.string.isRequired };
MissionItems.propTypes = { description: PropTypes.string.isRequired };
MissionItems.propTypes = { reserved: PropTypes.bool.isRequired };
