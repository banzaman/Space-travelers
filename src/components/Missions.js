import { useEffect } from 'react';
import { Badge, Button, Spinner } from 'react-bootstrap';
import { useDispatch, useSelector } from 'react-redux';
import { getMissions } from '../redux/missions/missionsSlice';
import '../assets/css/Mission.css';

export default function Missions() {
  const { missions, isLoading, error } = useSelector((state) => state.missions);
  const dispatch = useDispatch();
  useEffect(() => {
    dispatch(getMissions());
  }, [dispatch]);

  if (isLoading) {
    return (
      <Button variant="primary" disabled>
        <Spinner
          as="span"
          animation="grow"
          size="sm"
          role="status"
          aria-hidden="true"
        />
        Loading...
      </Button>
    );
  }
  if (error) {
    return (
      <Badge bg="danger">Something went wrong</Badge>
    );
  }

  return (
    <>
      <hr />
      <section className="mission-container">
        <div>
          <h5 className="mission-name">Mission</h5>
          <h5 className="description">Description</h5>
          <h5 className="status">Status</h5>
          <h5 className="join">{}</h5>
        </div>
        <ul>
          {missions.map((mission) => (
            <li key={mission.mission_id} className="mission-list">
              <h5 className="mission-name">{mission.mission_name}</h5>
              <p className="description">{mission.description}</p>
              <span className="status"><Badge bg="secondary">Not a member</Badge></span>
              <div className="join"><Button>Join Mission</Button></div>
            </li>
          ))}
        </ul>
      </section>
    </>
  );
}
