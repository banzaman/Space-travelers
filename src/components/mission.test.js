import renderer from 'react-test-renderer';
import { Provider } from 'react-redux';
import { BrowserRouter } from 'react-router-dom';
import MissionItems from './MissionItems';
import Missions from './Missions';
import MissionTest from './missionsReducer';
import store from '../redux/store';

const mission = {
  id: '2',
  name: 'Commercial Resupply Services',
  description: 'Commercial Resupply Services (CRS) are a series of contracts awarded by NASA from 2008â€“2016 for delivery of cargo.',
  reserved: false,
};

jest.mock('axios');
describe('Mission must pass the test', () => {
  test('added mission must return data', () => {
    expect(MissionTest.addedMissions()).toBeDefined();
  });
  test('added Mission return value length must be 4', () => {
    expect(MissionTest.addedMissions()).toHaveLength(3);
  });
  test('added Mission must return name t', () => {
    expect(MissionTest.addedMissions()[2].description).toBe('Amazing rocket!');
  });
});

describe('Mission component', () => {
  it('renders Mission component', () => {
    const rendering = renderer.create(
      <Provider store={store}>
        <BrowserRouter>
          <Missions />
        </BrowserRouter>
      </Provider>,
    ).toJSON();
    expect(rendering).toMatchSnapshot();
  });
  it('renders Mission component', () => {
    const itemMock = renderer.create(
      <Provider store={store}>
        <BrowserRouter>
          <MissionItems
            id={mission.id}
            name={mission.name}
            description={mission.description}
            reserved={mission.reserved}
          />
        </BrowserRouter>
      </Provider>,
    ).toJSON();
    expect(itemMock).toMatchSnapshot();
  });
});
