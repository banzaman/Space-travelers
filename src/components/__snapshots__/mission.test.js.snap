// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Mission component renders Mission component 1`] = `
<button
  className="btn btn-primary"
  disabled={true}
  type="button"
>
  <span
    aria-hidden="true"
    className="spinner-grow spinner-grow-sm"
    role="status"
  />
  Loading...
</button>
`;

exports[`Mission component renders Mission component 2`] = `
<li
  className="mission-list"
>
  <h5
    className="mission-name"
  >
    Commercial Resupply Services
  </h5>
  <p
    className="description"
  >
    Commercial Resupply Services (CRS) are a series of contracts awarded by NASA from 2008â€“2016 for delivery of cargo.
  </p>
  <span
    className="status"
  >
    <span
      className="badge bg-secondary"
    >
      NOT A MEMBER
    </span>
  </span>
  <div
    className="join"
  >
    <button
      className="btn btn-outline-secondary"
      disabled={false}
      onClick={[Function]}
      type="button"
    >
      Join Mission
    </button>
  </div>
</li>
`;
